pkg/zcapld/verify.go

```diff
@@ -0,0 +1,198 @@
+	if err != nil {
+		return fmt.Errorf("invalid capability chain: %w", err)
+	}
+
+	// 3. verify the invoker...
+	// authorized invoker must match the verification method itself OR
+	// the controller of the verification method
+	isInvoker, err := isInvoker(proof.Capability, c.VerificationMethod)
+	if err != nil {
+		return fmt.Errorf("isInvoke: %w", err)
+	}
+
+	if !isInvoker {
+		return errors.New("the authorized invoker does not match the verification method or its controller")
+	}
+
+	// Begin ControllerProofPurpose
+
+	// TODO the code here validates the proof's "created" time against an expected time:
+	//  nolint:lll // don't want to break the link in two
+	//  https://github.com/digitalbazaar/jsonld-signatures/blob/8d91bcb351702dde4863fab660d7ca1e5e90b2a2/lib/purposes/ProofPurpose.js#L49-L57.
+	//  Note: the higher layers in bedrock-edv-storage don't actually set `proof.created`, so `created` is always NaN.
+	//  Do we really need to verify the proof's date at this layer though? Isn't that the responsibility of a higher
+	//  layer, ie the one that parses and verifies the http signature?
+
+	// TODO verify authorization of verificationMethod.ID by controller for proof purpose `capabilityInvocation`.
+	//  Controller are probably DIDs. They have a "capabilityInvocation" property (just like DIDs) that has
+	//  verificationMethod IDs.
+
+	return nil
+}
+
+// nolint:funlen,gocyclo // TODO decompose verifyCapabilityChain into smaller units
+func (c *CapabilityInvocation) verifyCapabilityChain(
+	capability *Capability, intendedAction string, resolver CapabilityResolver) error {
+	// 1.1. Ensure `capabilityAction`, if given, is allowed; if the capability
+	// restricts the actions via `allowedAction` then it must be in the set.
+	if len(capability.AllowedAction) > 0 && intendedAction != "" &&
+		!stringsContain(capability.AllowedAction, intendedAction) {
+		return fmt.Errorf(
+			`capability action "%s" is not allowed by the capability; allowed actions are: %+v`,
+			intendedAction, capability.AllowedAction)
+	}
+
+	if c.ExpectedAction != intendedAction {
+		return fmt.Errorf(
+			`capability action "%s" does not match the expected capability action of "%s"`,
+			intendedAction, c.ExpectedAction)
+	}
+
+	// 3. Validate the capability delegation chain.
+	err := capability.ValidateCapabilityChain()
+	if err != nil {
+		return fmt.Errorf("invalid capability chain: %w", err)
+	}
+
+	// 2. Get the capability delegation chain for the capability.
+	capabilityChain, err := capability.CapabilityChain()
+	if err != nil {
+		return fmt.Errorf("failed to fetch capabilityChain: %w", err)
+	}
+
+	// 4. Verify root capability (note: it must *always* be dereferenced since
+	// it does not need to have a delegation proof to vouch for its authenticity
+	// ... dereferencing it prevents adversaries from submitting an invalid
+	// root capability that is accepted):
+	isRoot := len(capabilityChain) == 0
+
+	var rootURI string
+
+	switch isRoot {
```

Comment:
this is of cause a matter of style but I am just curious (for my future reference): is `switch` more readable in such scenarios than the regular `if`?

Reply:
Both are equally readable to me but I have a slight preference for `switch`.

Anyway, refactored as per your suggestion since I see both my PR reviewers are in agreement. :)